app/scripts/controllers/metametrics.js

```diff
@@ -479,13 +480,42 @@ export default class MetaMetricsController {
     }
   }
 
+  /**
+   * Method to track metrics event
+   *
+   * If an actionId is provided createEventFragment, finalizeEventFragment are used to track event,
+   * this is to achieve idempotent behaviour.
```

Comment:
"idempotent" seems like the wrong word here. This function still has a side effect, and there is no guarantee that calling it twice will have the same effect as calling it once.

I'm not sure exactly on why this is designed this way, but I'm guessing that the action is being tracked as a fragment so that we can effectively _buffer_ it over a certain period of time. If the same fragment gets added twice, they get deduplicated because of the unique action id (is this how fragments work? :thinking: ). This ensures that successive identical calls only result in a single metric event, as long as they are within the time window you've set here of one minute.

Does that sound right? We should attempt to explain this in the description, and maybe indicate the reason for the one minute time span. Presumably one minute was chosen so that it would be well within a service worker lifecycle, but long enough to allow for a reset even on a slow machine.

Reply:
I updated the comments here.

As a fragment is created actionId is `createEventFragment` check to see if a fragment with `actionId + uniqueIdentifier` combination already exists. If it exists new fragment is not created.

`trackEvent` method has been changed to use `createEventFragment` and `finalizeEventFragment` to ensure that duplicate events are not captured.

About 1 min delay you are right, I added this to comment.

Even, if service worker dies before event is finalised, event will still be there in store and I think will be finalised when controller revives, @brad-decker can you plz confirm.

Reply:
@brad-decker : a question is `uniqueIdentifier` unique for each fragment, in that case we actually do not need to compare actionId.

Reply:
it would be unique, except that it doesn't exist on every fragment. so we should still use actionId.

Fragments have two paths, the `success` path fires the `successEvent` when `finalizeFragment` is called without the `abandoned` property set to true. This *only happens from outside the controller*. When timeouts occur, or when the controller restarts, the events are finalized using the `failureEvent` which is meant for things like user rejections, or "abandoned cart" type scenarios.

So it seems we will have event loss if the service worker dies before the fragment is finalized. Honestly, though, I think the solution we discussed for storing events until they are fired to prevent event loss from the other PR to replace analytics-node would be a more ideal way of handling "idempotency" of trackEvent. Instead of using the fragment system we can just fire trackEvent, but add the actionId as a key (perhaps even make it our "messageId" we send to segment) and use that to deduplicate events in the storage mechanism before they are sent to segment.

Reply:
I agree, in fact I would prefer to close this PR and create another using that approach. We just need to insure that we generate a unique messageId and it will take care of idempotency. Here is the PR for segment events https://github.com/MetaMask/metamask-extension/pull/16198

Reply:
I will not see this task a blocker for release though.

Reply:
I started working on alternate approach here: https://github.com/MetaMask/metamask-extension/pull/16267